### ISSUE TITLE
Worker Terminate: worker process potentially disposed before terminate message

### ISSUE THREAD

**Issue by liliankasem** [Azure/azure-functions-host#8979]
Issue found in a [CRI](https://portal.microsofticm.com/imp/v3/incidents/details/354815015/home).

During this CRI investigation, I found some potential issues around the WorkerTerminate message leading to `Did not find any initialized language workers` errors.

A function invocation took more than 30 minutes and so the function timed out:

`Timeout value of 00:30:00 exceeded by function 'Functions.FUNCTIONNAME' (Id: '4ef2e849-46bc-42cf-8b4e-e23c748b96cc'). Initiating cancellation.`

Once the timeout occurred, a series of events followed:

```kusto
// EUS region
FunctionsLogs
| where TIMESTAMP >= datetime(2022-12-07T00:42:19.8486606Z) and TIMESTAMP <= datetime(2022-12-07T00:42:19.8623134Z)
| where AppName =~ 'APP_NAME' or AppName startswith 'APP_NAME'
| project-reorder TIMESTAMP, HostVersion, RoleInstance, FunctionName, Summary, FunctionInvocationId
| order by TIMESTAMP asc
```

> See CRI link for app name and more details

| Timestamp | Summary |
| ----------- | ---------- |
| 2022-12-07T00:42:19.8521687Z | Restarting channel with workerId: '8c103fa0-2e4c-424c-9cc6-f111540e391b' that is executing invocation: '4ef2e849-46bc-42cf-8b4e-e23c748b96cc' and timed out. |
| 2022-12-07T00:42:19.8559763Z | Attempting to dispose webhost or jobhost channel for workerId: '8c103fa0-2e4c-424c-9cc6-f111540e391b', runtime: 'dotnet-isolated' |
| 2022-12-07T00:42:19.8590489Z | Disposing language worker channel with id:8c103fa0-2e4c-424c-9cc6-f111540e391b |
| 2022-12-07T00:42:19.8623134Z |  Sending WorkerTerminate message with grace period 5 seconds. |

After the grace period message was sent, function executions start to fail with:

`Microsoft.Azure.WebJobs.Host.FunctionInvocationException : Exception while executing function: Functions.GetWalletAccountsCache ---> System.InvalidOperationException : Did not find any initialized language workers`

```kusto
// EUS region
FunctionsLogs
| where TIMESTAMP >= datetime(2022-12-07T00:42:19.8486606Z) and TIMESTAMP <= datetime(2022-12-07T00:43:19.8623134Z)
| where AppName =~ 'APP_NAME' or AppName startswith 'APP_NAME'
| where Level < 4
| project-reorder TIMESTAMP, HostVersion, RoleInstance, FunctionName, Summary, FunctionInvocationId
| order by TIMESTAMP asc
```

**Open questions:**

1. What is happening when the worker gets a worker terminate message?
2. Is the grace period being respected by the worker and the host?
3. It looks like we dispose the language worker before we send the worker terminate message which leads to language workers not being found
4. Is there a potential race condition here?

**Comment by liliankasem** [comment]
It looks like this is related to issue #8916 - a PR to fix the issue was merged and should be released in January. I will keep this issue till end of Jan to keep an eye on the issue.

**Comment by fabiocav** [comment]
Moving this to sprint 138 for validation once the host deployment is done.

**Comment by liliankasem** [comment]
Difficult to validate as their isn't  a single message we can look for that made it clear what was happening here; as this was a race condition issue, the exception we're seeing didn't happen every time. We're going to close this issue for now as we expect has been fixed in host versions 4.15+. If this occurs again please open a new issue.

**Comment by pober** [comment]
> Difficult to validate as their isn't a single message we can look for that made it clear what was happening here; as this was a race condition issue, the exception we're seeing didn't happen every time. We're going to close this issue for now as we expect has been fixed in host versions 4.15+. If this occurs again please open a new issue.

I'm not 100% sure if my problem is related to this issue, but since 2023-02-05 18:58 UTC some of our Function apps have been generating dosens of "Did not find any initialized language workers errors" errors.

Some details:
Location: South Central US
Runtime version: 4.15.1.19904
Stack: Node.js 16
App: prod-planyway-notification-app-secondary
HostInstanceId: 3f9e9b61-54a8-4c69-9077-d1c7d1070202
InvocationId:	fc01b5b3-0b9a-4c00-839c-6fe7545255da

Could you please advise if I should open a new issue or contact Azure support?

**Comment by liliankasem** [comment]
@pober I would suggest contacting azure support so they can investigate your specific case
